nav2-usability/scenario_code/lib/nested-routing/nested_routing_go_router_prototype.dart

```diff
@@ -0,0 +1,217 @@
+// Copyright 2022, the Flutter project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:go_router_prototype/go_router_prototype.dart';
+
+void main() {
+  runApp(BooksApp());
+}
+
+class BooksApp extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      routeInformationParser: _router.parser,
+      routerDelegate: _router.delegate,
+    );
+  }
+
+  late final _router = GoRouter(
+    routes: [
+      ShellRoute(
+        path: '/',
+        redirect: (routeState) async {
+          if (routeState.path == '/') {
+            return '/books/new';
+          }
+        },
+        builder: (context, child) {
+          late final int selectedIndex;
+          final childPath = RouteState.of(context).activeChild?.path;
+          if (childPath == null || childPath == 'books') {
+            selectedIndex = 0;
+          } else if (childPath == 'settings') {
+            selectedIndex = 1;
+          }
+
+          return AppScreen(
+            currentIndex: selectedIndex,
+            child: child,
+          );
+        },
+        routes: [
+          ShellRoute(
+            path: 'books',
+            redirect: (routeState) async {
+              if (routeState.path == '/books') {
+                return '/books/new';
+              }
+            },
+            builder: (context, child) => BooksScreen(
+              selectedTab:
+                  RouteState.of(context).activeChild!.path == 'new' ? 0 : 1,
```

Comment:
This logic is repeated on lines 31-37. Consider adding a helper function. I don't feel strongly, feel free to skip

Reply:
This is an interesting point - `RouteState.of(context)` returns a different object depending on the `BuildContext`. When it's called from the `/` route, the activeChild is either `books` or `settings`, but when it's called from the `/books` route it's either `new` or `all`.

This is new feature we added to support relative routes, so feedback is welcome on this.

Reply:
I'm going to leave this as-is, just for readability

Reply:
Sounds good. I agree readability should be prioritized here.

The `RouteState.of(context)` seems reasonable to me, it's very consistent with other APIs like `Theme.of(context)`!

... That said, I come from the .NET ecosystem where extension methods are heavily used. I'd be curious to learn more about the [style guide's recommendation against extension methods](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-extension). For example, we could have `context.routeState() or `context.theme()`. Regardless, that's a separate conversation than the go_router work :)

Reply:
I forgot that GoRouter actually has a `GoRouterHelper` class already, which adds some extension methods to `BuildContext`. Under-the-hood, they call `GoRouter.of(context)` so I imagine we'll keep both options. I didn't add extensions this to the prototype, though.